### PR TITLE
feat: Support loading manifest from artifacts for resource templates. Fixes #5878

### LINKED ISSUE #5878: Creating A Resource Using Git Artifact
## Summary 

I have a workflow with 2 steps. The first is simply doing a git clone in order to create a git artifact. The repository I clone has a workflow yaml in it. I then want to use the workflow yaml in the next step, where I have a resource template. That is, the resource template should create the workflow which was in the git repo cloned in step 1! 

## Diagnostics

What Kubernetes provider are you using? GKE

What version of Argo Workflows are you running? 3.0.1

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: possible-bug-
spec:
  entrypoint: all-steps
  templates:
  - name: all-steps
    steps:
    - - name: step-1-create-git-artifact
        template: git-clone 
    - - name: step-2-create-workflow
        template: create-resource
        arguments:
          parameters:
          - name: resource-manifest
            value: "steps.step-1-create-git-artifact.outputs.parameters.manifest"
  - name: git-clone
    inputs:
      artifacts:
      - name: argo-source
        path: /src
        git:
          repo: https://github.com/argoproj/argo-workflows.git
          revision: master
    container:
      image: golang:1.10
      workingDir: /src
      command: [sh, -c]
      args: ["sleep 1;
              cat examples/hello-world.yaml > /tmp/hello-world.yaml"]
    outputs:
      parameters:
        - name: manifest
          valueFrom:
            default: "hello-world.yaml didnt work"
            path: "/tmp/hello-world.yaml"
  - name: create-resource
    inputs:
      parameters:
      - name: resource-manifest
    resource:
      action: create
      manifest: "{{inputs.parameters.resource-manifest}}"
```

```
Paste the logs from the workflow controller:
 Error (exit code 1): error: error validating "/tmp/manifest.yaml": error validating data: invalid object to validate; if you choose to ignore these errors, turn validation off with --validate=false  
```

I have also seen this error when changing the yamls:

```
create-resource master-resource-creator-lhvgb-729962375  2s    Error (exit code 1): error: error parsing /tmp/manifest.yaml: error converting YAML to JSON: yaml: mapping values are not allowed in this context
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by book987**
Signed-off-by: book987 <book78987book@gmail.com>

Fixes #5878 

Continue work on #7263

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -282,8 +282,8 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 	pod.ObjectMeta.Annotations[common.AnnotationKeyDefaultContainer] = defaultContainer
 
 	// Add init container only if it needs input artifacts. This is also true for
-	// script templates (which needs to populate the script)
-	if len(tmpl.Inputs.Artifacts) > 0 || tmpl.GetType() == wfv1.TemplateTypeScript || woc.getContainerRuntimeExecutor() == common.ContainerRuntimeExecutorEmissary {
+	// resource and script templates (which needs to populate the script)
+	if len(tmpl.Inputs.Artifacts) > 0 || tmpl.GetType() == wfv1.TemplateTypeScript || tmpl.GetType() == wfv1.TemplateTypeResource || woc.getContainerRuntimeExecutor() == common.ContainerRuntimeExecutorEmissary {
```
Comment (alexec) [inline workflow/controller/workflowpod.go:286]: I think we should probably remove this check completely and always run the init container. But not in the PR. Tech debt.

File: workflow/executor/executor.go [hunk 2]
```diff
@@ -249,6 +249,9 @@ func (we *WorkflowExecutor) StageFiles() error {
 		filePath = common.ExecutorScriptSourcePath
 		body = []byte(we.Template.Script.Source)
 	case wfv1.TemplateTypeResource:
+		if we.Template.Resource.ManifestFrom != nil {
+			return nil
```
Comment (alexec) [inline workflow/executor/executor.go:253]: If path is empty maybe return an error?

File: workflow/executor/executor.go [hunk 3]
```diff
@@ -249,6 +249,9 @@ func (we *WorkflowExecutor) StageFiles() error {
 		filePath = common.ExecutorScriptSourcePath
 		body = []byte(we.Template.Script.Source)
 	case wfv1.TemplateTypeResource:
+		if we.Template.Resource.ManifestFrom != nil {
```
Comment (terrytangyuan) [inline workflow/executor/executor.go:252]: We should log something similar to what we did above:
```
	case wfv1.TemplateTypeScript:
		log.Infof("Loading script source to %s", common.ExecutorScriptSourcePath)
```

File: test/e2e/resource_template_test.go [hunk 4]
```diff
@@ -94,6 +94,52 @@ spec:
...
+                - name: argosay-container
+                  image: argoproj/argosay:v2
+                  command: ["/argosay"]
+                restartPolicy: Never
+      resource:
+        action: create
+        successCondition: status.phase == Succeeded
+        failureCondition: status.phase == Failed
+        manifestFrom:
+          path: /tmp/manifest.yaml
```
Comment (terrytangyuan) [inline test/e2e/resource_template_test.go:132]: I don't see `manifestFrom` being used anywhere in the implementation so I don't think path other than `/tmp/manifest.yaml` works.

**Review by terrytangyuan — CHANGES_REQUESTED** [review]

**Comment by alexec** [comment]
@book987 how's it going? Do you want any help or support?

**Comment by terrytangyuan** [comment]
Would you like to re-open this PR and fix the CI?

**Comment by book987** [comment]
> Would you like to re-open this PR and fix the CI?

Yeah but I don't know how to do this, could you help me? I've never done that.

**Comment by terrytangyuan** [comment]
Somehow I am not able to re-open this PR. @alexec Do you see the same?

**Comment by terrytangyuan** [comment]
Perhaps create a new PR then

**Comment by RoryDoherty** [comment]
@book987 if you're too busy to finish this then I would be happy to pick this up, Let me know?
I'm currently running into this issue myself and would like to see it fixed

**Comment by book987** [comment]
> @book987 if you're too busy to finish this then I would be happy to pick this up, Let me know? I'm currently running into this issue myself and would like to see it fixed

Hi I've just create a new PR #8657 with new change. Can I ask you to help review the code? CI has passed.

cc @terrytangyuan

**Pull request closed without merging**